executor/android/android_seccomp.h

```diff
@@ -102,3 +102,23 @@ static void set_app_seccomp_filter()
 	// Will fail() if anything fails.
 	install_filter(&f);
 }
+
+#if GOARCH_amd64 || GOARCH_386
+// In bionic mkdir calls mkdirat, there is no seccomp hole for mkdir.
```

Comment:
Interesting.
I think this needs a much larger comment, if I am reading this correctly there are lots of things happen here.
1. This is not due to bionic, right? This is because of glibc, right? If we would link with bionic, everything would work I think.
2. This is also due to static linking with glibc, right?
3. Why do we do this for amd64/386, but not for arm/arm64? I would assume all things are the same for these arches... Oh, I see, it's because there is no mkdir syscall on arm64 at all, so glibc also calls mkdirat. What about arm?
4. This will affect all sandboxes and execution on all Linux distros as well, even though it's in the android file. However, this will affect C repros only for android sandbox, so for Linux execution behavior may be different. Worth mentioning.
5. This relies on a previous forward declaration of these functions, right? These are extern "C", but we define them in C++.
6. We also rely on exact signature match (C++ has overloading!). If I change, say, const char* to char*, then it silently stops working. But if I change mode to short, it breaks compilation. I think this is somewhat fragile wrt libc headers (and this affects compilation/execution on any Linux). A more reliable way to do such override is:
```
int syzmkdir(const char* path, mode_t mode) __asm__("mkdir");
int syzmkdir(const char* path, mode_t mode) {...}
```
It is not subject to both silent breaking and breaking compilation due to slight argument type mismatch.

Reply:
1.  I'm afraid linking with bionic could cause other problems and you'd need to switch to clang I think?
2. This is due to static linking with glibc + seccomp.  There are no holes for these syscalls in Android because bionic doesn't need them.  I would prefer to not add more holes to our seccomp policy so as to reduce the risk of mis rating bugs.
3. This wasn't a problem for arm/arm64, when I tested this on amd64 in the sandbox in cuttlefish I found these problems.
4. Do you want me to move this out of the Android only #defs so that it's the same for all of linux?
5. No? This is a header file, so it gets included first before you call things.
6. Why doesn't this break for bionic then?  We compile C and C++ all the time against bionic and this never breaks?

Reply:
No need to link bionic, I am just asking to add a comment proportional to the amount of things that happen underneath these few simply looking lines.

Re 4, I dunno, let's leave it here for now. It has quite weird effects, it's android-specific, but affects everything, but only in executor, not in C repros.

Re 5, if you change the mode argument to short type, you will see how it's interconnection to the previous libc declaration.

Re 6, it's totally implementation-defined all-bets-are-off land. It's like relying on undefined behavior in C. The fact that it worked in one context does not mean anything. If libc on a particular distro has slighly different headers, it may either break compilation, or silently stop working (this is C++ with function overloading, so potentially we may have 2 overloads with subtly different types, one coming from libc, another is this, which one will be chosen is subject to all insane C++ rules re type conversions, priorities, viable sets, etc).
If we do the __asm__ trick, then we never compilation breakage and never get silent breakage. The code always calls mkdir, and we substitute the symbol.